photosynthesis_metrics/perceptual.py

```diff
@@ -0,0 +1,280 @@
+        self.std = std.view(1, 3, 1, 1)
+        
+        self.normalize_features = normalize_features
+        self.reduction = reduction
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Content loss between feature representations of prediction and target tensors.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        """
+        _validate_input(input_tensors=(prediction, target), allow_5d=False)
+        prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        # Normalize input
+        mean, std = self.mean.to(prediction), self.std.to(prediction)
+        prediction = (prediction - mean) / std
+        target = (target - mean) / std
+
+        self.model.to(prediction)
+        prediction_features = self.get_features(prediction)
+        target_features = self.get_features(target)
+        
+        distances = [self.distance(x, y) for x, y in zip(prediction_features, target_features)]
+
+        # Scale distances, then average in spatial dimensions, then stack and sum in channels dimension
+        loss = torch.cat([(d * w.to(d)).mean(dim=[2, 3]) for d, w in zip(distances, self.weights)], dim=1).sum(dim=1)
+
+        # Solve big memory consumption
+        torch.cuda.empty_cache()
+
+        if self.reduction == 'none':
+            return loss
+
+        return {'mean': loss.mean,
+                'sum': loss.sum
+                }[self.reduction](dim=0)
+
+    def get_features(self, x: torch.Tensor) -> List[torch.Tensor]:
+        """
+        Args:
+            x: torch.Tensor with shape (N, C, H, W)
+        
+        Returns:
+            features: List of features extracted from intermediate layers
+        """
+        features = []
+        for name, module in self.model.features._modules.items():
+            x = module(x)
+            if int(name) in self.layers:
+                features.append(self.normalize(x) if self.normalize_features else x)
+        return features
+
+    def normalize(self, x: torch.Tensor) -> torch.Tensor:
+        """Normalize feature maps in channel direction to unit length.
+        Args:
+            x: Tensor with shape (N, C, H, W)
+        Returns:
+            x_norm: Normalized input
+        """
+        norm_factor = torch.sqrt(torch.sum(x ** 2, dim=1, keepdim=True))
+        return x / (norm_factor + self.EPS)
+
+
+class StyleLoss(ContentLoss):
+    r"""Creates Style loss that can be used for image style transfer or as a measure in
+    image to image tasks.
+    Uses pretrained VGG models from torchvision. Features can be normalized before summation.
+    Expects input to be in range [0, 1] or normalized with ImageNet statistics into range [-1, 1]
```

Comment:
I think it is worth mentioning here that this is the loss that uses Gramm matrix computions